File: bin/acceptance/b21_mcp_e2e_fly.mjs
#!/usr/bin/env node
/**
 * B2.1 MCP e2e Fly — витрина → бариста (UI) → гость (WS).
 *   node bin/acceptance/b21_mcp_e2e_fly.mjs
 * Требует: tmp/b21_mcp_e2e_prep.json
 */
import { readFileSync, writeFileSync, mkdirSync } from "node:fs"
import { dirname, join } from "node:path"
import { fileURLToPath } from "node:url"
import { chromium } from "playwright"

const __dirname = dirname(fileURLToPath(import.meta.url))
const root = join(__dirname, "..", "..")
const prep = JSON.parse(readFileSync(join(root, "tmp/b21_mcp_e2e_prep.json"), "utf8"))
const outDir = join(root, prep.screen_dir)
const steps = []
const startedAt = new Date().toISOString()

const {
  base: BASE,
  barista_email: BARISTA_EMAIL,
  order_id: orderId,
  order_dom_id: orderDomId,
  reconnect_token: token,
  guest_base: guestBase,
  subtitles
} = prep

const PASSWORD = process.env.STAFF_PASSWORD || "demo123456"

function step(id, action, pass, extra = {}) {
  const row = { id, action, pass, at: new Date().toISOString(), ...extra }
  steps.push(row)
  console.log(`${pass ? "PASS" : "FAIL"} ${id}: ${action}`)
  return pass
}

mkdirSync(outDir, { recursive: true })

async function loginBarista(page) {
  await page.goto(`${BASE}/login`, { waitUntil: "networkidle", timeout: 60000 })
  const emailSel = 'input[name="email"], input[name="user[email]"]'
  const passSel = 'input[name="password"], input[name="user[password]"]'
  await page.locator(emailSel).first().fill(BARISTA_EMAIL)
  await page.locator(passSel).first().fill(PASSWORD)
  await page.locator('input[type="submit"], button[type="submit"]').first().click()
  await page.waitForURL(/\/(barista|manager|platform)/, { timeout: 30000 })
}

async function ensureShiftOpen(page) {
  await page.goto(`${BASE}/barista`, { waitUntil: "networkidle", timeout: 60000 })
  const body = await page.locator("body").innerText()
  if (body.includes("Смена открыта")) return true

  await page.goto(`${BASE}/barista/shift`, { waitUntil: "networkidle", timeout: 60000 })
  const openBtn = page.locator('input.btn-open-shift, button:has-text("Открыть смену")')
  if (await openBtn.count()) {
    await openBtn.click()
    await page.waitForURL(/\/barista/, { timeout: 30000 })
  }
  await page.goto(`${BASE}/barista`, { waitUntil: "networkidle" })
  return (await page.locator("body").innerText()).includes("Смена открыта")
}

async function openGuest(page) {
  await page.goto(guestBase, { waitUntil: "networkidle", timeout: 60000 })
  const reconnectOk = await page.evaluate(
    async ({ orderId, token, tid }) => {
      sessionStorage.setItem("shop_last_order_id", orderId)
      sessionStorage.setItem("shop_reconnect_token", token)
      const csrf = document.querySelector('meta[name="csrf-token"]')?.content || ""
      const apiKey = document.querySelector('meta[name="shop-api-key"]')?.content || ""
      const res = await fetch(`/shop/api/session/reconnect?tenant_id=${tid}`, {
        method: "POST",
        credentials: "same-origin",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          "X-CSRF-Token": csrf,
          "X-Shop-Api-Key": apiKey
        },
        body: JSON.stringify({ order_id: orderId, reconnect_token: token })
      })
      const body = await res.json().catch(() => ({}))
      return { ok: res.ok, status: body.status || body.ok }
    },
    { orderId, token, tid: prep.tenant_id }
  )
  await page.goto(`${guestBase}#/order/${orderId}`, { waitUntil: "networkidle", timeout: 60000 })
  await page.waitForSelector(".status-title, .error-text", { timeout: 30000 })
  return reconnectOk
}

const browser = await chromium.launch({ headless: true })
const guest = await browser.newPage({ viewport: { width: 390, height: 844 } })
const barista = await browser.newPage({ viewport: { width: 1280, height: 900 } })

let allPass = true

try {
  // Step 1 — vitrina
  await guest.goto(guestBase, { waitUntil: "networkidle", timeout: 60000 })
  const catalogOk = (await guest.locator("body").innerText()).length > 50
  await guest.screenshot({ path: join(outDir, "stage5_e2e_vitrina_checkout.png"), fullPage: true })
  step(
    "step1_vitrina_catalog",
    "GET /shop каталог + скрин stage5_e2e_vitrina_checkout.png",
    catalogOk,
    { screenshot: "stage5_e2e_vitrina_checkout.png" }
  )
  allPass &&= catalogOk

  const reconnectOk = await openGuest(guest)
  const acceptedTitle = await guest.locator(".status-title").innerText().catch(() => "")
  const orderAccepted =
    reconnectOk?.ok !== false &&
    (acceptedTitle.includes("Оплачен") || acceptedTitle.includes("Принят") || acceptedTitle.includes("Готовится"))
  step(
    "step1_vitrina_order_accepted",
    "Заказ accepted на экране гостя (prep API + OTP)",
    orderAccepted,
    { order_id: orderId, status_title: acceptedTitle }
  )
  allPass &&= orderAccepted

  // Step 2 — barista board
  await loginBarista(barista)
  const shiftOk = await ensureShiftOpen(barista)
  step("step2_shift_open", "Открытая смена на demo A", shiftOk)
  allPass &&= shiftOk

  await barista.goto(`${BASE}/barista`, { waitUntil: "networkidle" })
  const card = barista.locator(`#${orderDomId}`)
  await card.waitFor({ state: "visible", timeout: 30000 })
  const onBoard = await card.isVisible()
  await barista.screenshot({ path: join(outDir, "stage5_e2e_vitrina_to_board.png"), fullPage: true })
  step(
    "step2_barista_board",
    `Карточка ${orderDomId} в НОВЫЕ + скрин`,
    onBoard,
    { screenshot: "stage5_e2e_vitrina_to_board.png", order_dom_id: orderDomId }
  )
  allPass &&= onBoard

  // Step 3 — barista UI → preparing
  const prepBtn = card.locator('button.card-btn-status, input.card-btn-status').filter({ hasText: "ГОТОВИТСЯ" })
  await prepBtn.first().click()
  await barista.locator(`#orders-preparing #${orderDomId}`).waitFor({ state: "visible", timeout: 15000 })
  const inPreparing = await barista.locator(`#orders-preparing #${orderDomId}`).isVisible()
  await barista.screenshot({ path: join(outDir, "stage5_e2e_barista_preparing.png"), fullPage: true })
  step(
    "step3_barista_preparing_ui",
    "Клик ГОТОВИТСЯ → карточка в PREPARING (turbo)",
    inPreparing,
    { screenshot: "stage5_e2e_barista_preparing.png" }
  )
  allPass &&= inPreparing

  // Step 4 — guest preparing (WS ≤5s, no reload)
  const guestTitleBefore = await guest.locator(".status-title").innerText().catch(() => "")
  const wsStart = Date.now()
  let wsOk = false
  try {
    await guest.locator(".status-subtitle", { hasText: subtitles.preparing }).waitFor({ timeout: 5000 })
    wsOk = true
  } catch {
    wsOk = false
  }
  const wsMs = Date.now() - wsStart
  await guest.screenshot({ path: join(outDir, "stage3_guest_preparing.png"), fullPage: true })
  await guest.screenshot({ path: join(outDir, "stage5_e2e_guest_preparing.png"), fullPage: true })
  const prepSubtitle = await guest.locator(".status-subtitle").innerText().catch(() => "")
  step(
    "step4_guest_preparing_ws",
    "Подзаголовок «Ваш заказ начали готовить» без reload ≤5с",
    wsOk && prepSubtitle.includes(subtitles.preparing),
    {
      ws_ms: wsMs,
      title_before: guestTitleBefore,
      subtitle: prepSubtitle,
      screenshots: ["stage3_guest_preparing.png", "stage5_e2e_guest_preparing.png"]
    }
  )
  allPass &&= wsOk && prepSubtitle.includes(subtitles.preparing)

  // Step 5 — barista UI → ready
  const readyCard = barista.locator(`#orders-preparing #${orderDomId}`)
  const readyBtn = readyCard.locator('button.card-btn-status, input.card-btn-status').filter({ hasText: "ГОТОВ" })
  await readyBtn.first().click()
  await barista.locator(`#orders-ready #${orderDomId}`).waitFor({ state: "visible", timeout: 15000 })
  const inReady = await barista.locator(`#orders-ready #${orderDomId}`).isVisible()
  await barista.screenshot({ path: join(outDir, "stage5_e2e_barista_ready.png"), fullPage: true })
  step(
    "step5_barista_ready_ui",
    "Клик ГОТОВ → карточка в READY",
    inReady,
    { screenshot: "stage5_e2e_barista_ready.png" }
  )
  allPass &&= inReady

  // Step 6 — guest ready WS
  const ws2Start = Date.now()
  let ws2Ok = false
  try {
    await guest.locator(".status-subtitle", { hasText: subtitles.ready }).waitFor({ timeout: 5000 })
    ws2Ok = true
  } catch {
    ws2Ok = false
  }
  const ws2Ms = Date.now() - ws2Start
  await guest.screenshot({ path: join(outDir, "stage3_guest_ready.png"), fullPage: true })
  await guest.screenshot({ path: join(outDir, "stage5_e2e_guest_ready.png"), fullPage: true })
  const readySubtitle = await guest.locator(".status-subtitle").innerText().catch(() => "")
  step(
    "step6_guest_ready_ws",
    "«Заказ готов, забирайте!» без reload ≤5с",
    ws2Ok && readySubtitle.includes(subtitles.ready),
    {
      ws_ms: ws2Ms,
      subtitle: readySubtitle,
      screenshots: ["stage3_guest_ready.png", "stage5_e2e_guest_ready.png"]
    }
  )
  allPass &&= ws2Ok && readySubtitle.includes(subtitles.ready)
} catch (e) {
  step("step_error", e.message, false, { error: String(e) })
  allPass = false
} finally {
  await browser.close()
}

const result = {
  task: "B2_1_mcp_e2e_fly",
  tool: "playwright (MCP-style browser e2e)",
  stand: BASE,
  started_at: startedAt,
  finished_at: new Date().toISOString(),
  order_id: orderId,
  order_dom_id: orderDomId,
  tenant_id: prep.tenant_id,
  barista_email: BARISTA_EMAIL,
  screen_dir: prep.screen_dir,
  steps,
  status: allPass ? "PASS" : "FAIL"
}

writeFileSync(join(root, "tmp/b21_mcp_e2e_steps.json"), JSON.stringify(result, null, 2))
console.log(JSON.stringify(result, null, 2))
process.exit(allPass ? 0 : 1)

File: bin/acceptance/b21_mcp_e2e_fly.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# B2.1 MCP e2e Fly — полный прогон: prep + Playwright + smoke + тесты → JSON.
#   FLY_BIN=flyctl ruby bin/acceptance/b21_mcp_e2e_fly.rb

require "json"
require "open3"
require "fileutils"

ROOT = File.expand_path("../..", __dir__)
DATE = Time.now.utc.strftime("%Y-%m-%d")
OUT = ENV.fetch(
  "OUT",
  File.join(ROOT, "docs/operations/milestones/veha_2/artifacts/demo-feedback/b21_mcp_e2e_#{DATE}.json")
)
SCREEN_DIR = "docs/operations/milestones/veha_2/artifacts/demo-feedback/screenshots/b21_barista_board_2026-06-10"
FLY_BIN = ENV.fetch("FLY_BIN", "flyctl")

def run!(cmd, label)
  puts "==> #{label}: #{cmd}"
  system(cmd) || raise("#{label} failed")
end

def run_capture(cmd)
  out, status = Open3.capture2e(cmd)
  [ out, status.success? ]
end

checks = []
stand = {}

# --- Подготовка стенда ---
whoami, ok = run_capture("#{FLY_BIN} auth whoami 2>&1")
checks << { id: "flyctl_auth", pass: ok && !whoami.strip.empty?, whoami: whoami.strip }
raise "flyctl auth login required" unless checks.last[:pass]

stand = {
  base: ENV.fetch("BASE", "https://coffeeos.fly.dev"),
  tenant_id: ENV.fetch("B21_TENANT_ID", "2fdee1ac-4674-41ee-b89e-87b45643f789"),
  barista_email: ENV.fetch("B21_BARISTA_EMAIL", "barista-a@demo.coffeeos.local"),
  screen_dir: SCREEN_DIR,
  deploy_note: "пользователь: deploy develop выполнен"
}
checks << { id: "stand_constants", pass: true, **stand }

Dir.chdir(ROOT) do
  run!("FLY_BIN=#{FLY_BIN} ruby bin/acceptance/b21_mcp_e2e_prep.rb", "e2e prep")
  prep = JSON.parse(File.read("tmp/b21_mcp_e2e_prep.json"))
  checks << {
    id: "demo_shift_and_order",
    pass: prep["order_id"].to_s != "",
    order_id: prep["order_id"],
    order_dom_id: prep["order_dom_id"]
  }

  playwright_ok = system("node bin/acceptance/b21_mcp_e2e_fly.mjs")
  e2e_steps = JSON.parse(File.read("tmp/b21_mcp_e2e_steps.json"))
  checks << { id: "playwright_e2e", pass: playwright_ok && e2e_steps["status"] == "PASS", steps: e2e_steps["steps"] }

  smoke_ok = system("FLY_BIN=#{FLY_BIN} ruby bin/acceptance/b21_barista_board_fly_smoke.rb")
  smoke_path = Dir.glob("docs/operations/milestones/veha_2/artifacts/demo-feedback/b21_fly_smoke_*.json").max
  smoke = JSON.parse(File.read(smoke_path))
  checks << {
    id: "fly_smoke",
    pass: smoke_ok && smoke["status"] == "PASS",
    artifact: File.basename(smoke_path || "b21_fly_smoke_#{DATE}.json")
  }

  mcp_ok = system("ruby bin/acceptance/b21_mcp_fly_verify.rb")
  mcp_path = Dir.glob("docs/operations/milestones/veha_2/artifacts/demo-feedback/b21_mcp_fly_*.json").max
  mcp = JSON.parse(File.read(mcp_path))
  checks << { id: "mcp_fly_verify", pass: mcp_ok, status: mcp["status"], artifact: File.basename(mcp_path.to_s) }

  guest_ok = system("bin/rails test test/integration/b21_guest_notify_test.rb")
  checks << { id: "b21_guest_notify_test", pass: guest_ok }

  subset_ok = system(
    "bin/rails test test/services/barista/board_orders_query_test.rb " \
    "test/integration/barista_tablet_regression_test.rb:185"
  )
  checks << { id: "b21_board_subset", pass: subset_ok }

  screens = %w[
    stage5_e2e_vitrina_checkout.png stage5_e2e_vitrina_to_board.png
    stage5_e2e_barista_preparing.png stage5_e2e_barista_ready.png
    stage5_e2e_guest_preparing.png stage5_e2e_guest_ready.png
    stage3_guest_preparing.png stage3_guest_ready.png
  ].each_with_object({}) { |f, h| h[f] = File.exist?(File.join(SCREEN_DIR, f)) }
  checks << { id: "screenshots_on_disk", pass: screens.values.all?, files: screens }
end

status = checks.all? { |c| c[:pass] } ? "PASS" : "FAIL"

result = {
  scenario: "B2.1 MCP e2e — витрина → бариста (UI) → гость (WS)",
  task: "B2_1_mcp_e2e_fly",
  run_at: DATE,
  stand: stand,
  order_id: (JSON.parse(File.read(File.join(ROOT, "tmp/b21_mcp_e2e_prep.json"))) rescue {})["order_id"],
  e2e_steps: (JSON.parse(File.read(File.join(ROOT, "tmp/b21_mcp_e2e_steps.json"))) rescue {})["steps"],
  checks: checks,
  status: status,
  customer_signoff: false
}

File.write(OUT, JSON.pretty_generate(result))
puts JSON.pretty_generate(result)
puts "written #{OUT}"
exit(status == "PASS" ? 0 : 1)
